packages/kbn-health-gateway-server/src/kibana/routes/root.ts

```diff
@@ -45,62 +46,71 @@ export function createRootRoute({ config, log }: RootRouteDependencies) {
 }: {
   fetch: Fetch;
   kibanaConfig: KibanaConfig;
-  log: Logger;
+  logger: Logger;
 }) {
-  const requests = await Promise.allSettled(
+  const responses = await Promise.allSettled(
     kibanaConfig.hosts.map(async (host) => {
-      log.debug(`Fetching response from ${host}${KIBANA_ROOT_ROUTE}`);
+      logger.debug(`Fetching response from ${host}${KIBANA_ROOT_ROUTE}`);
       return fetch(`${host}${KIBANA_ROOT_ROUTE}`);
     })
   );
 
-  return requests.map((r, i) => {
-    if (r.status === 'rejected') {
-      log.error(`No response from ${kibanaConfig.hosts[i]}${KIBANA_ROOT_ROUTE}`);
-    } else {
-      log.info(
-        `Got response from ${kibanaConfig.hosts[i]}${KIBANA_ROOT_ROUTE}: ${JSON.stringify(
-          r.value.status
-        )}`
-      );
+  responses.forEach((response, index) => {
+    const host = `${kibanaConfig.hosts[index]}${KIBANA_ROOT_ROUTE}`;
+
+    if (response.status !== 'rejected') {
+      logger.debug(`Got response from ${host}: ${JSON.stringify(response.value.status)}`);
+
+      return;
+    }
+
+    if (response.reason instanceof Error) {
+      logger.error(response.reason);
+    }
+
+    if (response.reason instanceof Error && response.reason.name === 'AbortError') {
+      logger.error(`Request timeout for ${host}`);
+
+      return;
     }
-    return r;
+
+    logger.error(
+      `No response from ${host}: ${
+        response.reason instanceof Error ? response.reason.message : JSON.stringify(response.reason)
+      }`
+    );
   });
+
+  return responses;
 }
 
 function mergeResponses(
   responses: Array<PromiseFulfilledResult<Response> | PromiseRejectedResult>
 ) {
-  let statusCode = 200;
-  for (const response of responses) {
-    if (
-      response.status === 'rejected' ||
-      !isHealthyResponse(response.value.status, response.value.headers)
-    ) {
-      statusCode = 503;
-    }
-  }
+  const hasUnhealthyResponse = responses.some(isUnhealthyResponse);
 
   return {
     body: {}, // The control plane health check ignores the body, so we do the same
-    statusCode,
+    statusCode: hasUnhealthyResponse ? 503 : 200,
   };
 }
 
-function isHealthyResponse(statusCode: number, headers: Headers) {
-  return isSuccess(statusCode) || isUnauthorized(statusCode, headers);
+function isUnhealthyResponse(response: PromiseFulfilledResult<Response> | PromiseRejectedResult) {
+  return (
+    response.status === 'rejected' || !(isSuccess(response.value) || isUnauthorized(response.value))
+  );
 }
 
-function isUnauthorized(statusCode: number, headers: Headers): boolean {
-  return statusCode === 401 && headers.has('www-authenticate');
+function isUnauthorized({ status, headers }: Response): boolean {
+  return status === 401 && headers.has('www-authenticate');
 }
 
-function isSuccess(statusCode: number): boolean {
-  return (statusCode >= 200 && statusCode <= 299) || statusCode === 302;
+function isSuccess({ status }: Response): boolean {
+  return (status >= 200 && status <= 299) || status === 302;
```

Comment:
Unrelated to the PR, but why do we need to check for `302`'s here? Is the http client we're using (`node-fetch` apparently) not properly following redirections? @lukeelmers

Reply:
Yeah, following redirects was explicitly disabled in https://github.com/elastic/kibana/pull/144985. This was done for consistency with the current control plane behavior -- we learned recently that control plane's current health check actually doesn't follow redirects, so it just treats a `302` as healthy.

I guess this means that in practice the `isSuccess` check will probably always return true, and the `isUnauthorized` check will not really run, since the root `/` is gonna attempt to redirect to `/login` anyway.

Probably fine to keep as-is for completeness sake though. There's a slack thread on this I can point you to if you are curious.